sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1445,45 +1494,108 @@ _pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
+  return CurrentIndex;
+}
+
+// Return a queue descriptor using the provided ordinal and index.
+void _pi_queue::pi_queue_group_t::createQueueDesc(
+    uint32_t Index, uint32_t Ordinal,
+    ZeStruct<ze_command_queue_desc_t> &ZeCommandQueueDesc) {
+
+  ZeCommandQueueDesc.ordinal = Ordinal;
+  ZeCommandQueueDesc.index = Index;
   ZeCommandQueueDesc.mode = ZE_COMMAND_QUEUE_MODE_ASYNCHRONOUS;
 
   // Evaluate performance of explicit usage for "0" index.
-  if (ZeCommandQueueIndex != 0) {
+  if (Index != 0) {
     ZeCommandQueueDesc.flags = ZE_COMMAND_QUEUE_FLAG_EXPLICIT_ONLY;
   }
 
   zePrint("[getZeQueue]: create queue ordinal = %d, index = %d "
           "(round robin in [%d, %d])\n",
           ZeCommandQueueDesc.ordinal, ZeCommandQueueDesc.index, LowerIndex,
           UpperIndex);
+}
+
+// This function will return one of possibly multiple available native
+// queues and the value of the queue group ordinal.
+ze_command_queue_handle_t &
+_pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
+
+  // QueueIndex is the proper L0 index.
+  // Index is the plugins concept of index, with main and link copy engines in
+  // one range.
+  uint32_t QueueIndex;
+  auto Index = getQueueIndex(QueueGroupOrdinal, &QueueIndex);
+
+  ze_command_queue_handle_t &ZeQueue = ZeQueues[Index];
+  if (ZeQueue)
+    return ZeQueue;
+
+  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
+  createQueueDesc(QueueIndex, *QueueGroupOrdinal, ZeCommandQueueDesc);
 
   auto ZeResult = ZE_CALL_NOCHECK(
       zeCommandQueueCreate, (Queue->Context->ZeContext, Queue->Device->ZeDevice,
                              &ZeCommandQueueDesc, &ZeQueue));
   if (ZeResult) {
     die("[L0] getZeQueue: failed to create queue");
   }
+
   return ZeQueue;
 }
 
+// This function will return one of possibly multiple available
+// immediate commandlists associated with this Queue.
+pi_command_list_ptr_t &
+_pi_queue::pi_queue_group_t::getImmCmdList(bool UseCopyEngine) {
+
+  uint32_t QueueIndex, QueueOrdinal;
+  auto Index = Queue->getQueueGroup(UseCopyEngine)
+                   .getQueueIndex(&QueueOrdinal, &QueueIndex);
+
+  if (ImmCmdLists[Index] != Queue->CommandListMap.end())
+    return ImmCmdLists[Index];
+
+  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
+  createQueueDesc(QueueIndex, QueueOrdinal, ZeCommandQueueDesc);
+
+  ze_command_list_handle_t ZeCommandList;
+  ZE_CALL_NOCHECK(zeCommandListCreateImmediate,
+                  (Queue->Context->ZeContext, Queue->Device->ZeDevice,
+                   &ZeCommandQueueDesc, &ZeCommandList));
+  ImmCmdLists[Index] =
+      Queue->CommandListMap
+          .insert(std::pair<ze_command_list_handle_t, pi_command_list_info_t>{
+              ZeCommandList, {nullptr, false, nullptr, QueueOrdinal}})
+          .first;
+  // Add this commandlist to the cache so it can be destroyed as part of
+  // QueueRelease
+  auto QueueType = Type;
+  auto &ZeCommandListCache =
+      QueueType != queue_type::Compute
```

Comment:
Shouldn't the check be the opposite here?

Reply:
Yes.